### PR TITLE
Enhance Array to get age information of entries

### PR DESCRIPTION
**Pull request by furuame**

### CODE REVIEW

File: sparta/sparta/resources/Array.hpp [hunk 1]
```diff
@@ -884,6 +899,33 @@ namespace sparta
...
+        {
+            sparta_assert(ArrayT == ArrayType::AGED);
+
+            // Since our data_list_ always adds new items to the
+            // front.  The oldest data is actually kept at the end of
+            // the list.  We can iterate from the end to find the nth
+            // oldest item.
+            auto it = aged_list_.rbegin();
+            uint32_t idx = invalid_entry_;
+            for(uint32_t i = 0; i < aged_list_.size(); ++i)
```
Comment (knute-sifive) [inline sparta/sparta/resources/Array.hpp:918]: Since `aged_list_.size()` will not change in this loop, for debug builds, consider caching this value.

Comment (furuame) [reply sparta/sparta/resources/Array.hpp:918]: Ok

**Comment by knute-sifive** [comment]
So, the main purpose of this change is to make the age ordering number sequential?  A little lost on why there is a required update to the age numbering...

File: sparta/test/Array/Array_test.cpp [hunk 2]
```diff
@@ -433,6 +433,27 @@ int main()
...
+        EXPECT_EQUAL(aged_array_test.getAge(i), age_vec_1[idxx++]);
+    }
+    // getAge Test3
+    idxx = 0;
+    aged_array_test.erase(4); // Erase the last element (the oldest one)
+    std::vector<uint32_t> age_vec_2 {3, 0, 2, 1};
+    for(uint32_t i = 0; i < 4; ++i){
+        EXPECT_EQUAL(aged_array_test.getAge(i), age_vec_2[idxx++]);
+    }
+
```
[the remark this answers is not in the document]
Comment (furuame) [reply sparta/test/Array/Array_test.cpp:456]: Yes if you follow the original allocation order

[the remark this answers is not in the document]
Comment (knute-sifive) [reply sparta/test/Array/Array_test.cpp:456]: Mind adding a quick test for that -- ensure it starts at zero?

**Comment by furuame** [comment]
> So, the main purpose of this change is to make the age ordering number sequential? A little lost on why there is a required update to the age numbering...

The main purpose is to know the age of the specific entry. Yes they have to be sequential. If there was an array of 8 entries and we just popped 1 entry (say 6th oldest), we need to update 7th/8th (now they are actually 6th/7th)'s ages from 7/8 to 6/7.

File: sparta/test/Array/Array_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (furuame) [reply sparta/test/Array/Array_test.cpp:456]: Good idea

[the remark this answers is not in the document]
Comment (furuame) [reply sparta/test/Array/Array_test.cpp:469]: Done

**Review by knute-sifive — APPROVED** [review]
